sqlx-core/src/cursor.rs

```diff
@@ -0,0 +1,65 @@
+use std::future::Future;
+
+use futures_core::future::BoxFuture;
+use futures_core::stream::BoxStream;
+
+use crate::connection::MaybeOwnedConnection;
+use crate::database::{Database, HasRow};
+use crate::executor::Execute;
+use crate::{Connect, Pool, Row};
+
+/// Represents a result set, which is generated by executing a query against the database.
+///
+/// A `Cursor` can be created by either [`Executor::execute`](trait.Execute.html) or
+/// [`Query::fetch`](struct.Query.html).
+///
+/// Initially the `Cursor` is positioned before the first row. The `next` method moves the cursor
+/// to the next row, and because it returns `None` when there are no more rows, it can be used
+/// in a `while` loop to iterate through all returned rows.
+pub trait Cursor<'c, 'q, DB>
+where
+    Self: Send,
+    DB: Database,
+    // `.await`-ing a cursor will return the affected rows from the query
+    Self: Future<Output = crate::Result<u64>>,
+{
+    // Construct the [Cursor] from a [Pool]
+    // Meant for internal use only
+    // TODO: Anyone have any better ideas on how to instantiate cursors generically from a pool?
+    #[doc(hidden)]
+    fn from_pool<E>(pool: &Pool<DB::Connection>, query: E) -> Self
+    where
+        Self: Sized,
+        E: Execute<'q, DB>;
+
+    #[doc(hidden)]
+    fn from_connection<E, C>(conn: C, query: E) -> Self
+    where
+        Self: Sized,
+        DB::Connection: Connect,
+        // MaybeOwnedConnection<'c, DB::Connection>:
+        //     Connect<Database = DB>,
+        C: Into<MaybeOwnedConnection<'c, DB::Connection>>,
+        E: Execute<'q, DB>;
+
+    #[doc(hidden)]
+    fn first(self) -> BoxFuture<'c, crate::Result<Option<<DB as HasRow<'c>>::Row>>>
+    where
+        'q: 'c;
+
+    /// Fetch the next row in the result. Returns `None` if there are no more rows.
+    fn next(&mut self) -> BoxFuture<crate::Result<Option<<DB as HasRow>::Row>>>;
+
+    /// Map the `Row`s in this result to a different type, returning a [`Stream`] of the results.
+    fn map<'a, T, F>(self, f: F) -> BoxStream<'a, crate::Result<T>>
+    where
+        Self: Sized,
+        F: Send + Sync + 'static,
+        T: Send + Unpin + 'static,
+        F: for<'r> Fn(<DB as HasRow<'r>>::Row) -> T,
+    {
+        unimplemented!(
+            "Cursor::map is currently supplied by inherent methods to work around not having GATs"
+        )
+    }
+}
```

Comment:
Not a strong opinion by any means, but I wonder if it would be a good idea to use two traits here. One which is actually `pub(crate)` and then the other which is `pub`. Perhaps the `pub trait Cursor` here would just define the two actual pub items, and then something like `pub(crate) trait CursorInternal` for the internal `#[doc(hidden)]` bits.

Anyway, like I said, not a big deal. Thoughts?

Reply:
I believe @abonander had an idea to drop this method entirely and require users to do:

```rust
sqlx::query("SELECT ...")
  .bind(foo)
  .bind(bar)
  .map(|row| MyType { ... })  // map is here and on `Query` instead of `Cursor`
  .fetch_all(&mut conn)
  .await?;
```

_ideally_ we would have this method but really the cursor abstraction is meant to quite low level, once we get GATs we can make a massive amount of simplifications including just implementing `Stream` for `Cursor`.

Reply:
Thanks to @abonander and a ton of refactoring this has fallen out.